Add necessary patches for LLVM 3.7(.1)

Keno opened the pull request:
And make ORC JIT the default for those LLVM versions. This does
not activate the new LLVM versions by default yet, to give time
to update everything on CI, etc., but doing the actual activation
is a one line change after this.

@staticfloat @tkelman Once this is done, how do we go about
preparing our CI infrastructure for the switch?

Close https://github.com/JuliaLang/julia/issues/14466

Keno committed:
Add necessary patches for LLVM 3.7(.1)

And make ORC JIT the default for those LLVM versions. This does
not activate the new LLVM versions by default yet, to give time
to update everything on CI, etc., but doing the actual activation
is a one line change after this.

tkelman commented:
3.7.1 was already tagged but isn't up at http://llvm.org/releases/3.7.1 yet, so I think we should skip the 3.7.0 checksums. I have a test branch where now we can actually just build llvm from source and cache the result on Linux Docker Travis. Mac will need a homebrew bottle.

vtjnash commented on deps/llvm-3.7.1.patch:
```diff
@@ -0,0 +1,550 @@
+ 
+-  if (isa<TerminatorInst>(Inst) || isa<PHINode>(Inst))
++  if (isa<TerminatorInst>(Inst) || isa<PHINode>(Inst) ||
++      isa<LandingPadInst>(Inst))
```
why is this in the diff? we don't even use landingpads

Keno replied:
Miscompiles Cxx otherwise.

vtjnash commented on deps/llvm-3.7.1.patch:
```diff
@@ -0,0 +1,550 @@
++    NewFunc->setPersonalityFn(
++        MapValue(OldFunc->getPersonalityFn(), VMap,
++                 ModuleLevelChanges ? RF_None : RF_NoModuleLevelChanges,
++                 TypeMapper, Materializer));
```
why is this in the diff? we don't use personality functions either

Keno replied:
Same reason.

vtjnash commented on deps/llvm-3.7.1.patch:
```diff
@@ -0,0 +1,550 @@
+   AddrLabelSymbols = nullptr;
+ 
+   Context.reset();
++  FuncInfoMap.clear();
```
what's this? it alsy doesn't seem related to the SectionMM

Keno replied:
Persistence bug. I have all these patches with commit messages as the kf/patches branch on JuliaLang/llvm.

vtjnash commented on deps/llvm-3.7.1.patch:
```diff
@@ -0,0 +1,550 @@
+diff --git a/include/llvm/ExecutionEngine/SectionMemoryManager.h b/include/llvm/ExecutionEngine/SectionMemoryManager.h
+index 0b0dcb0..7bb96eb 100644
+--- a/include/llvm/ExecutionEngine/SectionMemoryManager.h
++++ b/include/llvm/ExecutionEngine/SectionMemoryManager.h
```
i'm not convinced that we want to maintain this as a patch (instead of temporarily using a private copy of this class)

Keno replied:
We'll still need the other patches, so that doesn't help much.

tkelman commented:
Need to ping llvm-dev about putting up source tarballs. I don't get what in their release process takes so long after tagging (http://lists.llvm.org/pipermail/llvm-dev/2015-December/093179.html was 10 days ago).

Keno commented:
Not sure what's going on with 3.7.1. Let's just go with 3.7.0 for now and switch later.

tkelman commented:
It's more of a time investment for me to rebuild binaries across 2-3 different platforms than it is to get the llvm release manager to finish the job here. I'm not subscribed to llvm-dev, can we ask what the holdup is here?

hayd commented:
This thread was started yesterday: http://lists.llvm.org/pipermail/llvm-dev/2015-December/093572.html

Keno commented:
It was suggested that we switch LLVM to git-external to be able to pull in the release branch even though the tarballs aren't up yet. What would we do about the patches? Those are based on the LLVM version which we wouldn't know any more.

ViralBShah commented:
Can't we just create a source tarball of 3.7.1 on our own and host it on s3? If llvm.org releases a tarball, we can switch to it.

Keno commented:
Yeah, I'm in favor of that solution as well. Let's just do the switch. It's up to @tkelman though. I'm happy to do whatever is required to finally get this done.

tkelman commented:
With a git external we'd be using sha instead of version number, so the patch application should be tweaked to also check for that.

Using a different url for one specific version is pretty messy, but maybe the fastest way to get something temporary in place.

ViralBShah commented:
Why is it so messy? It just changes a URL for download that can be reverted when the official tarball is made available. Seems like the smallest change to get this going.

Keno committed:
Mirror LLVM 3.7.1 sources on S3

Keno commented:
Ok, S3 mirror is up. Should be good to go. @tkelman shall we merge this, so you can prepare the buildbots and then we'll do the actual switch in a separate PR?

tkelman commented:
I think you need to fix the permissions so that bucket is readable. What are the chances your tarballs will end up identical to the real ones once they go up?

Keno commented:
I thought the bucket was readable. I tested with this branch, but that was from within AWS, so maybe that was the difference.

Keno commented:
Ok, try again. The files were public, but I hadn't made the bucket listable.

Keno commented:
Oh, and I think the content is the same (this is the final tags), but I don't know how the release tarballs are built so I'm pretty sure they won't be bit-identical.

tkelman commented:
we should make the directory name inside the tarball consistent with upstream releases, so llvm-3.7.1.src

tkelman commented:
The release win32 build, but not the debug build, segfaults at the start of bootstrap on this branch with llvm 3.7.1.

Keno commented:
win64 is fine?

tkelman commented:
will let you know when it finishes

Keno commented:
Also, what gcc version? Anything > 4.9 miscompiles llvm.

tkelman commented:
4.9.2. That's a bit of a problem, given clang as a mingw-w64 cross compiler isn't especially easy to set up or use. Can we patch llvm or add a gcc flag to work around it? Fedora and Arch are also on gcc 5 by now afaik.

Keno commented:
Yes, you can use `-fno-ipa-ra`.

tkelman commented:
What would be the downside of adding that flag anywhere we're using gcc? Is it needed just for llvm, or libjulia, or (somehow) all deps?

Keno commented:
Probably fine to add it unconditionally when targeting mingw. Do we have a way to detect compiler version?

tkelman commented:
We need > 4.7 due to c++11 in new llvm anyway. Looks like there's a proposed patch for the gcc issue at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66655, just need to build mingw to test it.

win64 actually looks okay, tests are passing and performance looks normal. getting `<invalid>` in backtraces if I try to start with `--precompiled=yes` though.

Keno commented:
Ok, I'll take a look at the win32 failure.

tkelman commented:
remember that you need a 32 bit gdb to debug i686 mingw applications... i keep forgetting that one

Keno commented:
How do I get a 32bit gdb. I can reproduce the segfault now.

tkelman commented:
Either from 32 bit cygwin, or I think some of the mingw-builds come with a native gdb.

Keno commented:
ok

tkelman commented:
looks like opensuse also has a mingw32-gdb package, so adding that to the get-toolchain script might work too. segfault looks like it's in attrbuilder, but you're probably just about at the same backtrace now.

Keno commented:
Yep, I got it.

Keno commented:
Ok, the problem seems to be a disagreement on what the correct stack boundary should be

tkelman commented:
That sounds really familiar. We use a -mincoming-stack-boundary flag for openblas, fftw, and libgit2.

Keno commented:
Yeah, I tried that option, but it doesn't work. Now trying `-mstackrealign`.

Keno commented:
Doesn't work either. Not sure what the issue is.

Keno commented:
Ah, I was an idiot and only did it for C code. Patch coming.

Keno committed:
Set incoming stack boundary for libjulia on win32

Keno commented:
Still seeing a segfault unfortunately, but it appears to be a different issue.

Keno commented:
Seems to fail in the stack guard. Will continue investigating tomorrow.

tkelman commented:
Hm looks like `-fno-ipa-ra` is an error for gcc 4? Ick. I'm building mingw gcc with that patch on the opensuse build service, will see if it fixes the reduced test case reported on gcc's bugzilla. If so I can ask the cygwin mingw maintainer to apply the same patch.

tkelman commented:
Someone with a mac is going to need to prepare a llvm37-julia homebrew formula and submit it to https://github.com/staticfloat/homebrew-julia so @staticfloat can bottle it

Keno commented:
Fundamental remaining bug seems to be broken codegen for getfield:
```
function is_assigned_inner(sv::StaticVarInfo, s::Symbol)
    ccall(:jl_,Void,(Any,),sv)
    ccall(:jl_,Void,(Any,),s)
    ccall(:jl_,Void,(Any,),sv.vinfo)
    for vi in sv.vinfo
        ccall(:jl_,Void,(Any,),vi)
        if vi[1] === s
            return (vi[3]&4) != 0
        end
    end
    return false
end
```
```
Core.Inference.StaticVarInfo(sp=svec(), cenv=Core.Inference.ObjectIdDict(ht=Array{Any, 1}[#<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>]), vars=Array{Any, 1}[:e], gensym_types=Array{Any, 1}[], vinfo=Array{Any, 1}[Array{Any, 1}[:e, Any, 0], Array{Any, 1}[:##args#533, Tuple{Symbol, Any}, 0]], label_counter=0, fedbackvars=Core.Inference.ObjectIdDict(ht=Array{Any, 1}[#<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>]))
:e
Array{Any, 1}[:e]
:e
A method error occured before the base module was defined. Aborting...
getindex
(:e, 1)
```
Notice that `.vinfo` results in the value of `vars` instead.

ViralBShah commented:
@staticfloat Would you have some time to prepare it? If not, I can prepare it. Are there instructions somewhere?

Keno commented:
LLVM IR looks correct: https://gist.github.com/Keno/dc60023d2bd31eeac6bc

Keno commented:
Seems to be an IR optimization issue. I wonder if the FPM is not getting the correct data layout.

Keno commented:
Yep, I think that's it. Yay!

Keno committed:
Clean up module initialization

Make jl_setup_module idempotent and remove a bunch of ad-hoc code
that was supposed fix up the module after creating the execution
engine.

In the process, fix a bug where not all modules would have the
layout set, causing mis-optimizations.

Keno commented:
Ok, getting there. Will add the fallback for 3.3 that's causing CI to fail. Lastly, `-O` fails. Probably the same initialization bug I fixed on 3.8 (though it doesn't exist in the form I fixed it, but maybe a similar one).

tkelman commented:
This'll need a bit of squashing when it's ready. And the folder name inside the tarballs would still be good to fix.

staticfloat commented:
@ViralBShah the basic way of doing this is going to be looking at [the current julia LLVM 3.3 formula](https://github.com/staticfloat/homebrew-julia/blob/master/llvm33-julia.rb), or possibly [the current mainline homebrew LLVM 3.6.1 formula](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/llvm.rb), changing what needs to be changed in order to build LLVM 3.7.1, building it, and verifying that it works with Julia.  That will also require a bit of tweaking of the [julia formula](https://github.com/staticfloat/homebrew-julia/blob/master/julia.rb), just to change things from `llvm33-julia` to `llvm37-julia` or somesuch.  (I suggest making a `llvm37-julia.rb` file to contain the new LLVM formula in the `homebrew-julia` tap)

tkelman commented:
also check homebrew-versions, I think they may have a formula for 3.7.0 or an open PR for one

tkelman commented:
and given https://github.com/JuliaLang/julia/commit/a9b3015bf5553f141187c160ead4cd2723d9cee0, may want to manually

tkelman commented:
sorry wrong button. given https://github.com/JuliaLang/julia/commit/a9b3015bf5553f141187c160ead4cd2723d9cee0#diff-9, may want to manually define USE_ORCJIT if we know our homebrew llvm is patched

Keno commented:
I'm suspecting the remaining win32 failure to be another miscompilation that's not fixed by the indicated workaround. I'm actually quite speechless as to how bad gcc has gotten in version 5. I think we should move ahead anyway, and just tell people not to use `-O` on win32 for now. Maybe we can switch our build system to clang later.

tkelman commented:
4.9.2 from cygwin actually looks like it's working now. Which toolchain were you using for win32, opensuse or the one from msys2 pacman? What kind of failure are you seeing?

I used the opensuse build service to apply the suggested patch from the gcc bug thread and it fixed the test case, so hopefully that may get applied before gcc 6 and/or 5.4. Could also direct the `get_toolchain.sh` script at my repository to see if it helps. If it does, I could submit the patch to the opensuse cross-compiler maintainers.

Keno commented:
msys2 pacman gcc 5.3.0. I'm seeing LLVM being miscompiled. Easily checked by running the llvm tests to see if they're failing.

tkelman commented:
"our build system" does not use msys2 pacman's gcc. On win32 that's also dwarf exception handling, so I'm surprised it worked at all. I do still see
```
     * subarray             LLVM ERROR: Unable to allocate section memory!
Makefile:12: recipe for target 'all' failed
make[1]: *** [all] Error 1
Makefile:573: recipe for target 'testall1' failed
make: *** [testall1] Error 2
```
which looks similar to #11083.

LLVM tests don't run particularly easily here, `make -C deps check-llvm` is a no-op in cross compiles.

Keno committed:
Make it easier to build with clang on mingw

Keno commented:
I realized I was wrong about my gcc. It's the one we get using `contrib/windows/get_toolchain.sh`, i.e. the opensuse one. It seems not only does gcc 5.3 miscompile LLVM, but it's libstdc++ is also broken (hopefully to be fixed by 5.4 though), so I can't even compile with clang at the moment.

Keno commented:
@tkelman I can't reproduce the subarray error.

tkelman commented:
try running tests in a single process with `make testall1`

ViralBShah commented:
@staticfloat @tkelman This seems to require diving into a fair amount of homebrew, which I am happy to do, but won't be able to for all of the coming week.

Keno commented:
I can believe that `make testall1` would run out of memory. However, I'm not entirely convinced there's too much to do about that. One thing we could do is leave pages as RWX such that we can avoid internal fragmentation, but having RWX pages is not really best practice. I am a little surprised that virtual memory usage is still so high though. Not entirely sure why that is.

tkelman commented:
It isn't broken now and it is broken under llvm 3.7, so it's definitely a regression.

Keno commented:
Ok, I can try to do to improve the situation, but I don't think we should hold up the transition on other platforms because of it.

tkelman commented:
squash out commits that fail to build, fix the folder name in the tarballs (and adjust the checksums accordingly) then I think we can merge this and move to the next step.

Keno commented:
Done

tkelman commented:
aaand they finally put up the real release today at http://llvm.org/releases/3.7.1. Ping me when homebrew stuff is getting ready and I'll do a PR to switch the CI on all 5 main platforms.
